### ISSUE TITLE
Add `tuple_permutations`

### ISSUE THREAD

**Issue by JonathanBrouwer** [rust-itertools/itertools#1001]
There is the `combinations()` and `tuple_combinations()` functions, I'm however missing the `tuple_permutations()` function.

Is there a reason this couldn't exist?
If not, I might even be open to implementing it depending on how much macro hackery is needed

**Comment by ronnodas** [comment]
It may be easier to add `array_permutations()` if you want to avoid macro hackery.

**Comment by JonathanBrouwer** [comment]
That's true and in general probably nicer to use than tuples, thanks for the pointer!

**Comment by ronnodas** [comment]
It seems possible to share most of the code from `permutations()`, similar to the `array_combinations()` implementation. In that case we probably want to start with:
```rust
impl<T> PoolIndex<T> for Box<[usize]> {
    type Item = Vec<T>;

   ...
}
```
and then make `PermutationState` generic. (Aside: I'm not sure why `PermutationState::Loaded.indices` is a `Box<[usize]>` instead of a `Vec<usize>`.)

**Comment by phimuemue** [comment]
> Aside: I'm not sure why `PermutationState::Loaded.indices` is a `Box<[usize]>` instead of a `Vec<usize>`.

Canonically we use `Box<[usize]>` if the size stays unchanged.
